### PR TITLE
Optionally generate source maps.

### PR DESCRIPTION
**Pull request by mprobst**
This doesn't quite work yet - for some reason Chrome doesn't pick up the source map.

### CODE REVIEW

**Comment by mprobst** [comment]
@alexeagle PTAL, I think this is good.

File: lib/main.ts [hunk 1]
```diff
@@ -1,22 +1,45 @@
...
   lastCommentIdx: number = -1;
-  currentFile: ts.SourceFile;
-  relativeFileName: string;
+
+  failFast: boolean;
   errors: string[] = [];
 
-  constructor(private failFast: boolean = false, private generateLibraryName: boolean = false) {}
+  constructor(opts: TranspilerOptions = {}) {
+    this.failFast = !!opts.failFast;
```
Comment (alexeagle) [inline lib/main.ts]: why do we store individual fields in the class, rather than a single TranspilerOptions reference?

File: lib/main.ts [hunk 2]
```diff
@@ -26,11 +49,18 @@ export class Transpiler {
...
   translateProgram(program: ts.Program, relativeFileName: string): string {
     this.relativeFileName = relativeFileName;
-    return program.getSourceFiles()
-        .filter((sourceFile: ts.SourceFile) => !sourceFile.fileName.match(/\.d\.ts$/) &&
-                                               !!sourceFile.fileName.match(/\.[jt]s$/))
-        .map((f) => this.translate(f))
-        .join('\n');
+    var src = program.getSourceFiles()
+                  .filter((sourceFile: ts.SourceFile) => !sourceFile.fileName.match(/\.d\.ts$/) &&
+                                                         !!sourceFile.fileName.match(/\.[jt]s$/))
```
Comment (alexeagle) [inline lib/main.ts]: is this really clang-formatted?

File: lib/main.ts [hunk 3]
```diff
@@ -26,11 +49,18 @@ export class Transpiler {
...
+                  .map((f) => this.translate(f))
+                  .join('\n');
+    if (this.sourceMap) src += this.generateSourceMapComment();
+    return src;
+  }
+
+  generateSourceMapComment() {
+    var base64map = new Buffer(JSON.stringify(this.sourceMap)).toString('base64');
+    return '\n\n//# sourceMappingURL=data:application/json;base64,' + base64map;
```
Comment (alexeagle) [inline lib/main.ts]: nit: template string rather than concatenation?

File: lib/main.ts [hunk 4]
```diff
@@ -80,19 +110,51 @@ export class Transpiler {
...
       var e = new Error(this.errors.join('\n'));
       e.name = 'TS2DartError';
       throw e;
     }
+
     return this.result;
   }
 
+  enterNode(n: ts.Node) {
+    if (!this.sourceMap) return; // source maps disabled.
```
Comment (alexeagle) [inline lib/main.ts]: if this method is only for sourcemap generation, maybe it should have a different name?

File: lib/main.ts [hunk 5]
```diff
@@ -80,19 +110,51 @@ export class Transpiler {
...
+    var file = n.getSourceFile() || this.currentFile;
+    var start = n.getStart(file);
+    var pos = file.getLineAndCharacterOfPosition(start);
+
+    var mapping: SourceMap.Mapping = {
+      original: {line: pos.line + 1, column: pos.character},
```
Comment (alexeagle) [inline lib/main.ts]: doesn't the character need a +1 also?

File: lib/main.ts (hunk 2 above)
Comment (mprobst) [inline lib/main.ts]: This part is, it aligns the boolean clauses with the start of expression at `=>`. The source in createCompilerHost below sadly causes clang-format it to completely drop the ball. I'll investigate.

File: lib/main.ts [hunk 6]
```diff
@@ -80,19 +110,51 @@ export class Transpiler {
...
+      generated: {line: this.line, column: this.column},
```
Comment (alexeagle) [inline lib/main.ts]: this means the source map only references the unformatted code, huh? maybe that's what you were trying to tell me yesterday :)

File: lib/main.ts (hunk 1 above)
Comment (mprobst) [inline lib/main.ts]: No strong reason - I thought it was clearer to pull out the fields and coerce them into booleans here. I can change if you feel strongly about it.

File: lib/main.ts [hunk 7]
```diff
@@ -773,7 +836,7 @@ export class Transpiler {
         this.emit(`'''${this.escapeTextForTemplateString(node)}'''`);
         break;
       case ts.SyntaxKind.TemplateMiddle:
-        this.result += this.escapeTextForTemplateString(node);
+        this.emitNoSpace(this.escapeTextForTemplateString(node));
```
Comment (alexeagle) [inline lib/main.ts]: these changes hint at a refactoring: maybe the result should be a private field in another class which represents the generated code. that class could encapsulate the responsibility of tracking the line and column

File: lib/main.ts (hunk 3 above)
Comment (mprobst) [inline lib/main.ts]: YMMV, but for simple appending to a string I find this more readable.

**Comment by alexeagle** [comment]
nice, LGTM
I guess you resolved the conflict between generating the sourcemap for dart vs. our code having sourcemap back to TS at runtime?

File: lib/main.ts (hunk 4 above)
Comment (mprobst) [inline lib/main.ts]: Good point, done.

File: lib/main.ts (hunk 5 above)
Comment (mprobst) [inline lib/main.ts]: I've experimentally established that it doesn't. It's odd, but until we get bug reports I'm willing to go with this :-)

File: lib/main.ts (hunk 6 above)
Comment (mprobst) [inline lib/main.ts]: Yes :-)

File: lib/main.ts (hunk 7 above)
Comment (mprobst) [inline lib/main.ts]: Done.

**Comment by mprobst** [comment]
I've sent https://github.com/evanw/node-source-map-support/pull/48 for the source map issue.

**Pull request merged**